multicluster/controllers/multicluster/core/remote_cluster_manager.go

```diff
@@ -0,0 +1,169 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"sync"
+
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+)
+
+type clusterEvent struct {
+	isAdd         bool
+	remoteCluster RemoteCluster
+}
+
+type RemoteClusterManager interface {
+	// Start RemoteClusterManager on an event loop in a go routine
+	Start() error
+	// Stop RemoteClusterManager
+	Stop() error
+	// AddRemoteCluster adds a remote cluster to RemoteClusterManager.
+	AddRemoteCluster(remoteCluster RemoteCluster)
+	// RemoveRemoteCLuster removes a remote cluster from RemoteClusterManager.
+	RemoveRemoteCluster(remoteCluster RemoteCluster)
+	// GetRemteClusters returns all remote clusters
+	GetRemoteClusters() map[common.ClusterID]RemoteCluster
+	// GetElectedLeaderClusterID returns the leader remote cluster or INVALID_CLUSTER_ID if none elected
+	GetElectedLeaderClusterID() common.ClusterID
+	// GetLocalClusterID returns local cluster ID
+	GetLocalClusterID() common.ClusterID
+}
+
+// The manager manages all leaders defined in the cluster set and manages common area access into them.
+// It also interfaces with leader election to know the elected-leader from which resources
+// must be monitored.
+type remoteClusterManager struct {
+	clusterSetID common.ClusterSetID
+
+	// map of all remote clusters to which we have access to their common area
+	remoteClusters map[common.ClusterID]RemoteCluster
+
+	// an elected leader in which certain operations need to be performed
+	electedLeaderCluster RemoteCluster
+
+	// set to true if we require leader election
+	needElection bool
+
+	// Local cluster ID
+	clusterID common.ClusterID
+
+	eventChan chan clusterEvent
+
+	// Copy of remoteClusters for access from cluster set reconciler to avoid locking on remoteClusters
+	clusterSyncMap sync.Map
+
+	stopFunc context.CancelFunc
+}
+
+func NewRemoteClusterManager(clusterSetID common.ClusterSetID, clusterID common.ClusterID) RemoteClusterManager {
+	return &remoteClusterManager{
+		clusterSetID:   clusterSetID,
+		clusterID:      clusterID,
+		eventChan:      make(chan clusterEvent),
+		remoteClusters: make(map[common.ClusterID]RemoteCluster),
+	}
+}
+
+/**
+ * RemoteClusterManager implementation
+ */
+
+func (r *remoteClusterManager) Start() error {
+	klog.Infof("Starting remote cluster manager for", "clusterset", r.clusterSetID)
+	stopCtx, stopFunc := context.WithCancel(context.Background())
+	r.StartLeaderElection()
+
+	go func() {
```

Comment:
We should not need another goroutine, when Start() is already in a goroutine, or maybe we should change not to run Start() in a goroutine.

Reply:
Sure, I can remove the goroutine here.

Reply:
Done, pushed a new commit to keep this change separate so it is easy to review, please let me know if it looks better.